packages/rollup-contracts/contracts/utils/MerkleTrieLib.sol

```diff
@@ -0,0 +1,171 @@
+        RLPReader.RLPItem[] decoded;
+    }
+
+
+    /*
+     * Internal Functions
+     */
+
+    /**
+     * @notice Checks a trie inclusion proof.
+     * @param _key Key of the node to verify.
+     * @param _value Value of the node to verify.
+     * @param _root Root of the trie.
+     * @param _proof Encoded proof.
+     * @return `true` if the node is in the trie, `false` otherwise.
+     */
+    function verifyInclusionProof(
+        bytes memory _key,
+        bytes memory _value,
+        bytes32 _root,
+        bytes memory _proof
+    ) public pure returns (bool) {
+        RLPReader.RLPItem[] memory proof = RLPReader.toList(RLPReader.toRlpItem(_proof));
+
+        // Convert the key into a series of half-packed bytes.
+        bytes memory key = _key.toNibbles();
+
+        bytes32 currentHash = _root;
+        uint256 keyIndex = 0;
+        for (uint256 i = 0; i < proof.length; i++) {
+            ProofElement memory node = ProofElement({
+                encoded: RLPReader.toBytes(proof[i]),
+                decoded: RLPReader.toList(RLPReader.toRlpItem(RLPReader.toBytes(proof[i])))
+            });
+
+            if (keyIndex == 0) {
+                // First proof element is always the root node.
+                require(
+                    keccak256(node.encoded) == currentHash,
+                    "Invalid root hash"
+                );
+            } else if (node.encoded.length >= 32) {
+                // Nodes 32 bytes or larger are hashed inside branch nodes.
+                require(
+                    keccak256(node.encoded) == currentHash,
+                    "Invalid large internal hash"
+                );
+            } else {
+                // Nodes smaller than 31 bytes aren't hashed.
+                require(
+                    node.encoded.toBytes32() == currentHash,
+                    "Invalid internal node hash"
+                );
+            }
+
+            // Nodes with `TREE_RADIX + 1` elements are branches.
+            if (node.decoded.length == TREE_RADIX + 1) {
+                if (keyIndex >= key.length) {
+                    // Value may sometimes be included at a branch node.
+                    return (
+                        RLPReader.toBytes(node.decoded[node.decoded.length-1]).equal(_value)
+                    );
+                } else {
+                    // Find the next node within the branch node and repeat.
+                    RLPReader.RLPItem memory next = node.decoded[uint8(key[keyIndex])];
+                    currentHash = getCorrectBytes(next).toBytes32();
+                    keyIndex++;
+                    continue;
+                }
+            }
+
+            // Nodes with two elements are either leaves or extensions.
+            if (node.decoded.length == 2) {
+                // Throw this step into a new function to avoid `STACK_TOO_DEEP`.
+                bool done;
+                (done, currentHash, keyIndex) = checkNonBranchNode(
```

Comment:
Looks like `keyIndex` will only ever increase if `!done`, and always be `0` when `done`. Does it make sense to remove `done` and convey done that way, (maybe with a `DONE_KEY_INDEX` const?) or is that too cryptic?

Reply:
Hmmmm. Returning `0` for `keyIndex` was mainly a way to deal with the fact that `checkNonBranchNode` must return all three values. It might be a little confusing to make this change, could make it seem like `keyIndex == 0` is a necessary requirement for being done.

Reply:
Sounds good! I thought it may be a stretch but thought I'd mention. I support leaving it in 👍